src/volatile_memory.rs

```diff
@@ -671,6 +679,223 @@ impl<'a, T: ByteValued> VolatileRef<'a, T> {
+    /// let vslice = mem_ref.get_slice(0, 32).map_err(|_| ())?;
+    /// let mut buf = [5u8; 16];
+    /// vslice.copy_to(&mut buf[..]);
+    /// for v in &buf[..] {
+    ///     assert_eq!(buf[0], 0);
+    /// }
+    /// # Ok(())
+    /// # }
+    /// ```
+    pub fn copy_to(&self, buf: &mut [T]) -> usize {
+        let mut addr = self.addr;
+        let mut i = 0;
+        for v in buf.iter_mut().take(self.len()) {
+            unsafe {
+                *v = read_volatile(addr as *const Packed<T>).0;
+                addr = addr.add(self.element_size());
+            };
+            i += 1;
+        }
+        i
+    }
+
+    /// Copies `self.len()` or `slice.len()` bytes, whichever is smaller, to `slice`.
+    ///
+    /// The copies happen in an undefined order.
+    /// # Examples
+    ///
+    /// ```
+    /// # use vm_memory::VolatileMemory;
+    /// # fn test_write_null() -> Result<(), ()> {
+    /// let mut mem = [0u8; 32];
+    /// let mem_ref = &mut mem[..];
+    /// let vslice = mem_ref.get_slice(0, 32).map_err(|_| ())?;
+    /// vslice.copy_to_volatile_slice(vslice.get_slice(16, 16).map_err(|_| ())?);
+    /// # Ok(())
+    /// # }
+    /// ```
+    pub fn copy_to_volatile_slice(&self, slice: VolatileSlice) {
+        unsafe {
+            // Safe because the pointers are range-checked when the slices
+            // are created, and they never escape the VolatileSlices.
+            // FIXME: ... however, is it really okay to mix non-volatile
+            // operations such as copy with read_volatile and write_volatile?
+            copy(
+                self.addr,
+                slice.addr,
+                min(self.len() * self.element_size(), slice.size),
+            );
+        }
+    }
+
+    /// Copies `self.len()` or `buf.len()` times the size of `T` bytes, whichever is smaller, to
+    /// this slice's memory.
+    ///
+    /// The copy happens from smallest to largest address in `T` sized chunks using volatile writes.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// # use std::fs::File;
+    /// # use std::path::Path;
+    /// # use vm_memory::VolatileMemory;
+    /// # fn test_write_null() -> Result<(), ()> {
+    /// let mut mem = [0u8; 32];
+    /// let mem_ref = &mut mem[..];
+    /// let vslice = mem_ref.get_slice(0, 32).map_err(|_| ())?;
+    /// let buf = [5u8; 64];
+    /// vslice.copy_from(&buf[..]);
+    /// for i in 0..4 {
+    ///     assert_eq!(vslice.get_ref::<u32>(i * 4).map_err(|_| ())?.load(), 0x05050505);
+    /// }
+    /// # Ok(())
+    /// # }
+    /// ```
+    pub fn copy_from(&self, buf: &[T]) {
+        let mut addr = self.addr;
+        for &v in buf.iter().take(self.len()) {
+            unsafe {
+                // Safe because the pointers are range-checked when the slices
+                // are created, and they never escape the VolatileSlices.
+                write_volatile(addr as *mut Packed<T>, Packed::<T>(v));
+                addr = addr.add(self.element_size());
```

Comment:
Again with the `ptr::add` and overflowing an `isize` :-s

Reply:
Here there is an implicit check against `self.len()` (thanks to `.take(self.len())`) so it's enough to check that the size fits an `isize` in get_array_ref.